### PR TITLE
Support Mod with fcode

### LINKED ISSUE #15230: autowrap fortran fails with expression containing Mod function twice
Minimal example:

```
from sympy import Matrix, Mod,, symbols
from sympy.utilities.autowrap import autowrap

x, y = symbols(['x', 'y'])
expr = Mod(x, 2)- Mod(y, 2)
f = autowrap(expr, args=[x, y])
```

Error:
```
CodeWrapError: Error while executing command: /home/hen/miniconda/envs/py37/bin/python -c import numpy.f2py as f2py2e;f2py2e.main() -c -m wrapper_module_5 wrapped_code_5.f90. Command output is:
running build
running config_cc
unifing config_cc, config, build_clib, build_ext, build commands --compiler options
running config_fc
unifing config_fc, config, build_clib, build_ext, build commands --fcompiler options
running build_src
build_src
building extension "wrapper_module_5" sources
f2py options: []
f2py:> /tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5module.c
creating /tmp/tmpz9l25tyj/src.linux-x86_64-3.7
Reading fortran codes...
	Reading file 'wrapped_code_5.f90' (format:free)
Post-processing...
	Block: wrapper_module_5
			Block: autofunc
Post-processing (stage 2)...
Building modules...
	Building module "wrapper_module_5"...
		Creating wrapper for Fortran function "autofunc"("autofunc")...
		Constructing wrapper function "autofunc"...
		  autofunc = autofunc(x,y)
	Wrote C/API module "wrapper_module_5" to file "/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5module.c"
	Fortran 77 wrappers are saved to "/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5-f2pywrappers.f"
  adding '/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/fortranobject.c' to sources.
  adding '/tmp/tmpz9l25tyj/src.linux-x86_64-3.7' to include_dirs.
copying /home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/f2py/src/fortranobject.c -> /tmp/tmpz9l25tyj/src.linux-x86_64-3.7
copying /home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/f2py/src/fortranobject.h -> /tmp/tmpz9l25tyj/src.linux-x86_64-3.7
  adding '/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5-f2pywrappers.f' to sources.
build_src: building npy-pkg config files
running build_ext
customize UnixCCompiler
customize UnixCCompiler using build_ext
get_default_fcompiler: matching types: '['gnu95', 'intel', 'lahey', 'pg', 'absoft', 'nag', 'vast', 'compaq', 'intele', 'intelem', 'gnu', 'g95', 'pathf95', 'nagfor']'
customize Gnu95FCompiler
Found executable /usr/bin/gfortran
customize Gnu95FCompiler
customize Gnu95FCompiler using build_ext
building 'wrapper_module_5' extension
compiling C sources
C compiler: gcc -pthread -B /home/hen/miniconda/envs/py37/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC

creating /tmp/tmpz9l25tyj/tmp
creating /tmp/tmpz9l25tyj/tmp/tmpz9l25tyj
creating /tmp/tmpz9l25tyj/tmp/tmpz9l25tyj/src.linux-x86_64-3.7
compile options: '-I/tmp/tmpz9l25tyj/src.linux-x86_64-3.7 -I/home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include -I/home/hen/miniconda/envs/py37/include/python3.7m -c'
gcc: /tmp/tmpz9l25tyj/src.linux-x86_64-3.7/fortranobject.c
gcc: /tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5module.c
In file included from /home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1823,
                 from /home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:18,
                 from /home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from /tmp/tmpz9l25tyj/src.linux-x86_64-3.7/fortranobject.h:13,
                 from /tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5module.c:15:
/home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
 #warning "Using deprecated NumPy API, disable it by " \
  ^~~~~~~
/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5module.c: In function ‘PyInit_wrapper_module_5’:
/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5module.c:333:3: warning: this ‘for’ clause does not guard... [-Wmisleading-indentation]
   for(i=0;f2py_routine_defs[i].name!=NULL;i++)
   ^~~
/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5module.c:336:5: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘for’
     {
     ^
In file included from /home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1823,
                 from /home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:18,
                 from /home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from /tmp/tmpz9l25tyj/src.linux-x86_64-3.7/fortranobject.h:13,
                 from /tmp/tmpz9l25tyj/src.linux-x86_64-3.7/fortranobject.c:2:
/home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
 #warning "Using deprecated NumPy API, disable it by " \
  ^~~~~~~
/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/fortranobject.c: In function ‘format_def’:
/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/fortranobject.c:138:18: warning: comparison of integer expressions of different signedness: ‘Py_ssize_t’ {aka ‘long int’} and ‘long unsigned int’ [-Wsign-compare]
         if (size < sizeof(notalloc)) {
                  ^
compiling Fortran sources
Fortran f77 compiler: /usr/bin/gfortran -Wall -g -ffixed-form -fno-second-underscore -fPIC -O3 -funroll-loops
Fortran f90 compiler: /usr/bin/gfortran -Wall -g -fno-second-underscore -fPIC -O3 -funroll-loops
Fortran fix compiler: /usr/bin/gfortran -Wall -g -ffixed-form -fno-second-underscore -Wall -g -fno-second-underscore -fPIC -O3 -funroll-loops
compile options: '-I/tmp/tmpz9l25tyj/src.linux-x86_64-3.7 -I/home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include -I/home/hen/miniconda/envs/py37/include/python3.7m -c'
gfortran:f90: wrapped_code_5.f90
error: Command "/usr/bin/gfortran -Wall -g -fno-second-underscore -fPIC -O3 -funroll-loops -I/tmp/tmpz9l25tyj/src.linux-x86_64-3.7 -I/home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include -I/home/hen/miniconda/envs/py37/include/python3.7m -c -c wrapped_code_5.f90 -o /tmp/tmpz9l25tyj/wrapped_code_5.o" failed with exit status 1
wrapped_code_5.f90:6:13:

 REAL*8 :: Mod
             1
Error: Symbol ‘mod’ at (1) already has basic type of REAL
wrapped_code_5.f90:6:13:

 REAL*8 :: Mod
             1
Error: Symbol ‘mod’ at (1) already has basic type of REAL
```

And here's the generated code.
```
!******************************************************************************
!*                       Code generated with sympy 1.2                        *
!*                                                                            *
!*              See http://www.sympy.org/ for more information.               *
!*                                                                            *
!*                      This file is part of 'autowrap'                       *
!******************************************************************************

REAL*8 function autofunc(x, y)
implicit none
REAL*8, intent(in) :: x
REAL*8, intent(in) :: y

REAL*8 :: Mod
REAL*8 :: Mod
autofunc = Mod(1.0*x, 2.0) - Mod(1.0*y, 2.0)

end function
```

Clearly the problem is that Mod gets imported twice. Hopefully this is an easy fix but I don't know where to start.

### PR DESCRIPTION
**Pull request by moorepants**
#### References to other Issues or PRs

Fixes #15230 

#### Brief description of what is fixed or changed

Added `Mod: mod` to the list of known Fortran functions.

#### Release Notes

* printing
  * Added Mod to the supported Fortran functions.

### CODE REVIEW

**Comment by bjodah** [comment]
My fortran is a bit rusty, so take the following with a grain of salt:

Doesn't `mod` convert arguments to integers? (I see that the test uses floating point, just a bit worried that this might be compiler specific?). I thought `dmod` is for (double precision) floating point numbers. I guess we could use `.is_integer` in `_print_Mod`?

If standard >=95 I think the printer can use `modulo` which is type-generic (?).

**Comment by moorepants** [comment]
I'm quite ignorant of Fortran. Just went with obvious, but potentially incorrect, solution. Should probably think more carefully about it. I noticed the c printer had some logic wrt to types.

**Comment by hsharrison** [comment]
FWIW the current behavior is correct for an expression with a single mod.

```
expr = x % 2.1
f = autowrap(expr, args=[x])
assert np.isclose(f(2.5), 0.4)
```

Which I think is just writing it as `Mod`. Fortran is case insensitive, right? So think just plain `mod` should work.

I haven't had a chance to test out this branch yet on my full code but I'll do so tomorrow morning.

**Comment by asmeurer** [comment]
@certik may also know what the correct usage is here.

**Comment by moorepants** [comment]
This https://www.obliquity.com/computer/fortran/intrinsic.html says that mod in Fortran 77 takes either Doulbe, Reals, or Integers.

**Comment by moorepants** [comment]
This https://docs.oracle.com/cd/E19422-01/819-3684/3_F77_Intrins.html breaks it down into different functions:
selection_267

**Comment by moorepants** [comment]
More: http://www.public.iastate.edu/~fortran/decfortran-axp/11,67.html:

```
MOD (dividend, divisor)

 A function that divides the first argument by the second and
 returns the remainder.
 +------+---------+----------+------------+-------------+        
 | Args | Generic | Specific |  Argument  | Result Type |
 +------+---------+----------+------------+-------------+
 |  2   |  MOD    |  --      | INTEGER*1  | INTEGER*1   |
 |      |         | IMOD     | INTEGER*2  | INTEGER*2   |
 |      |         | JMOD     | INTEGER*4  | INTEGER*4   |
 |      |see note | KMOD     | INTEGER*8  | INTEGER*8   |
 |      |         | AMOD     | REAL*4     | REAL*4      |
 |      |         | DMOD     | REAL*8     | REAL*8      |
 |      |see note | QMOD     | REAL*16    | REAL*16     |
 +------+---------+----------+------------+-------------+

 NOTE: KMOD is only available on AXP systems.
       QMOD is not available in DEC Fortran.
```

**Comment by certik** [comment]
Fortran has two functions, `mod` and `modulo`. Citing from [here](http://computer-programming-forum.com/49-fortran/8841595d70c918ba.htm):

> MOD is defined to be:
>
>   MOD(A,B) = A - int(A/B) * B
>
> while MODULO is defined as
>
>   MODULO(A,B) = A - floor(A/B) * B 

They both correctly work for real or integer.

**Comment by asmeurer** [comment]
The difference is that in MOD the sign is the same as A and in MODULO the sign is the same as B. I guess we should default to MODULO since Python's % and SymPy's `Mod` both exhibit that behavior.

**Comment by bjodah** [comment]
I think modulo requires fortran 90. But maybe compilers support it for fortran 77 as well?

**Comment by asmeurer** [comment]
Would there be any reason to use the specific over the generic? Is MODULO supported by all the Fortran standards we support?

**Comment by moorepants** [comment]
I'd try things out but of course my conda gfortran seems to be broken: https://github.com/conda-forge/libgfortran-feedstock/issues/15

**Comment by certik** [comment]
Yes, `modulo` is supported in all Fortran compilers (I just tried GFortran, Intel and NAG).

**Comment by moorepants** [comment]
`fcode.py` says it prints Fortran 77 code, yet we compile with gfortran in autowrap which doesn't do 77 by default, right?

**Comment by asmeurer** [comment]
But if someone picks a lower standard in the code generation, it means they only want that standard. So we should give a "not supported in the standard" error if MODULO isn't in the specified standard.

**Comment by certik** [comment]
All current Fortran compilers support at least F90, so I suggest we support features that can be compiled with any Fortran compiler in the last 5 years, let's say.

**Comment by asmeurer** [comment]
The printer has a `standard` flag that supports several standards: https://github.com/sympy/sympy/blob/4f41286b22a05d5d75f456f1e574d8c115bab5d4/sympy/printing/fcode.py#L132

The default is 77: https://github.com/sympy/sympy/blob/4f41286b22a05d5d75f456f1e574d8c115bab5d4/sympy/printing/fcode.py#L104

We can debate what the default should be (ideally in another issue). I'm not really clear why 77 is the default if most compilers support 90. But regardless, the printer should respect the standard given if MODULO isn't supported and raise an exception (I guess NotImplementedError is used in other places).

**Comment by certik** [comment]
Yes, we can do `NotImplementedError` for 66 and 77. And then in another issue, I suggest to remove 66 and 77.

**Comment by asmeurer** [comment]
I don't know if we should remove them, unless they are a large burden to support. At the very least we should try to get some feedback from people before doing so. There could be folks who have to support 77- or 66-only codebases.

**Comment by moorepants** [comment]
Ok, this has been updated to reflect the conversation. It is working locally for me. Ready for review.

**Review by certik — APPROVED** [review]
That looks good to me. Thanks @moorepants!

File: sympy/external/tests/test_autowrap.py [hunk 1]
```diff
@@ -142,6 +142,22 @@ def runtest_issue_10274(language, backend):
...
+    x, y = symbols('x, y')
+    expr = Mod(x, 3.0) - Mod(y, -2.0)
+    f = autowrap(expr, args=[x, y], language='F95')
+    exp_res = float(expr.xreplace({x: 3.5, y: 2.7}).evalf())
+    assert abs(f(3.5, 2.7) - exp_res) < 1e-14
+
+    x, y = symbols('x, y', integer=True)
+    expr = Mod(x, 3) - Mod(y, -2)
+    f = autowrap(expr, args=[x, y], language='F95')
+    exp_res = expr.xreplace({x: 3, y: 2}).evalf()
```
Comment (moorepants) [inline sympy/external/tests/test_autowrap.py]: This probably shouldn't evalf() so it stays an integer?

File: sympy/printing/fcode.py [hunk 2]
```diff
@@ -304,6 +304,19 @@ def _print_Function(self, expr):
...
 
+    def _print_Mod(self, expr):
+        # NOTE : Fortran has the functions mod() and modulo(). modulo() behaves
+        # the same wrt to the sign of the arguments as Python and SymPy's
+        # modulus computations (% and Mod()) but is not available in Fortran 66
+        # or Fortran 77, thus we raise an error.
+        if self._settings['standard'] in [66, 77]:
+            msg = ("Python % operator and SymPy's Mod() function are not "
+                   "supported by Fortran 66 or 77 standards.")
+            raise ValueError(msg)
```
Comment (moorepants) [inline sympy/printing/fcode.py]: Is NotImplementedError preferred?

File: sympy/utilities/codegen.py [hunk 3]
```diff
@@ -1204,7 +1204,7 @@ def _call_printer(self, routine):
                 assign_to = result.result_var
 
             constants, not_fortran, f_expr = self._printer_method_with_settings(
-                'doprint', dict(human=False, source_format='free'),
+                'doprint', dict(human=False, source_format='free', standard=95),
```
Comment (moorepants) [inline sympy/utilities/codegen.py]: It is not clear to me if the f2py call enforces the use of the F95 compiler. Anybody know?

File: sympy/printing/fcode.py (hunk 2 above)
Comment (certik) [reply sympy/printing/fcode.py]: I think it is.

**Comment by certik** [comment]
+1 to the last commit.

**Comment by moorepants** [comment]
Ok, tests passing. This can be merged.

**Pull request merged**